### PR TITLE
Improve input classifiers

### PR DESCRIPTION
**Pull request by GioSensation**
**Reviewer:** @alistairjcbrown 
**Asana:** https://app.asana.com/0/0/1201353015477166/f

## Description
This includes a bunch of improvements to input classifiers by attempting to read the text surrounding the field when explicit labels are not defined.

## Steps to test
Automated tests cover this.

### CODE REVIEW

File: src/Form/Form.js [hunk 1]
```diff
@@ -243,7 +243,7 @@ class Form {
 
     autofillEmail (alias, dataType = 'emailNew') {
         this.execOnInputs(
-            (input) => !input.matches(PASSWORD_SELECTOR) && this.autofillInput(input, alias, dataType),
```
Comment (GioSensation) [inline src/Form/Form.js:246]: Now we're passing `dataType`, so we no longer need this conditional.

File: src/Form/input-classifiers.js [hunk 2]
```diff
@@ -207,49 +182,39 @@ const getInputSubtype = (input) =>
...
-        match = label.textContent.match(regex)
-        if (match) return match
-    }
-
-    return null
-}
+const matchInPlaceholderAndLabels = (input, regex, form) =>
+    input.placeholder?.match(regex) ||
+    getExplicitLabelsText(input).match(regex) ||
+    getRelatedText(input, form).match(regex)
```
Comment (GioSensation) [inline src/Form/input-classifiers.js:195]: Simplified 🧘.

File: src/Form/input-classifiers.test.js [hunk 3]
```diff
@@ -79,17 +116,17 @@ describe('Input Classifiers', () => {
...
-            { text: '__-____', expectedResult: '12-2025' },
-            { text: 'mm-yy', expectedResult: '12-25' },
-            { text: 'i.e. 10-2022', expectedResult: '12-2025' },
-            { text: 'MM-AAAA', expectedResult: '12-2025' },
-            { text: 'mm_jj', expectedResult: '12_25' },
-            { text: 'mm.yy', expectedResult: '12.25' },
-            { text: 'mm - yy', expectedResult: '12-25' },
-            { text: 'mm yy', expectedResult: '12 25' },
-            { text: 'ie: 08.22', expectedResult: '12.25' }
+            { text: 'mm-yyyy', expectedResult: '08-2025' },
```
Comment (GioSensation) [inline src/Form/input-classifiers.test.js:119]: I switched to a single-digit month so we can test that the month is properly padded.

File: src/Form/selectors.js [hunk 4]
```diff
@@ -76,35 +100,52 @@ const CC_EXP_SELECTOR = `
...
+        negativeRegex: /mm[/\s.\-_—–]/i
     },
-    [CC_EXP_SELECTOR]: {
-        ccType: 'expiration',
-        regex: /exp(iry|iration)?/i
+    {
+        type: 'expiration',
+        selector: CC_EXP_SELECTOR,
+        regex: /(mm|\d\d)[/\s.\-_—–](yy|jj|aa|\d\d)|exp|valid/i,
+        negativeRegex: /invalid/i
```
Comment (GioSensation) [inline src/Form/selectors.js:144]: Safari and Webkit don't support negative lookbehinds, so I had to use `negativeRegex`.

File: src/autofill-utils.js [hunk 5]
```diff
@@ -92,6 +93,7 @@ const setValueForSelect = (el, val) => {
             // Events fire on the select el, not option
             events.forEach((ev) => el.dispatchEvent(ev))
             option.selected = true
+            events.forEach((ev) => el.dispatchEvent(ev))
```
Comment (GioSensation) [inline src/autofill-utils.js:96]: Some forms do stuff after the field is filled (like extra formatting). By firing the events again we make sure everything works as expected.

**Review by alistairjcbrown — APPROVED** [review]
Added a question and couple of non-blocking comments -- LGTM :shipit:

File: src/Form/Form.js (hunk 1 above)
Comment (GioSensation) [reply src/Form/Form.js:246]: Yep, correct.

File: src/Form/input-classifiers.js [hunk 6]
```diff
@@ -18,100 +15,85 @@ const testAgainstRegexes = (string = '', regex, negativeRegex) =>
...
+ * @type (el: HTMLInputElement, form: HTMLFormElement) => Boolean
  */
 const isPassword = (el, form) =>
-    checkMatch({
-        el,
-        form,
-        selector: PASSWORD_SELECTOR,
-        regex: /password/i
-    })
+    checkMatch({el, form, ...PASSWORD_MATCHER})
```
[the remark this answers is not in the document]
Comment (GioSensation) [reply src/Form/input-classifiers.js:78]: None 🙂. When I wrote the signature I didn't have those `matcher` objects, they were all separate parameters which was confusing. Now that we have packaged that info in an object I can use your suggestion. 👍

**Pull request merged**